# this script will annotate cellranger atac peaks with the fantom enhancer db and evaluate
# Cicero coaccessibility scores between promoter-enhancer connections

library(openxlsx)
library(dplyr)
library(here)
library(tibble)
library(Signac)
library(stringr)
library(annotatr)
library(GenomicRanges)
library(plyranges)
library(data.table)
library(EnsDb.Hsapiens.v86)
library(Seurat)

# read in cicero connections generated by ciceroFindConns.R
Read_Cicero_Conns <- function(clusterID) {
  connsFile <- paste0("analysis_control/ccans/monocle2/ciceroConns.control.",clusterID,".csv")
  filtered_cicero_conns <- fread(connsFile) %>%
    dplyr::select(Peak1, Peak2, coaccess) %>%
    dplyr::filter(coaccess > 0.1)
  return(filtered_cicero_conns)
}  

Prom_Enh_Anno <- function(clusterID, anno.db) {
  print(paste0("Annotating: ", clusterID))
  anno=anno.db
  
  # read in cicero ccan
  cicero_conns <- Read_Cicero_Conns(clusterID)
  cicero_conns$connID <- seq(nrow(cicero_conns))
  
  # concert cicero conns to granges and annotate the peaks
  cicero_peak1 <- str_split(cicero_conns$Peak1, pattern="_", simplify = TRUE) %>%
    as.data.frame()
  colnames(cicero_peak1) <- c("chr","start","end")
  cicero_peak1.gr <- makeGRangesFromDataFrame(cicero_peak1)
  cicero_peak1.gr$connID <- cicero_conns$connID
  peak1_anno <- annotate_regions(cicero_peak1.gr, annotations=anno, ignore.strand=TRUE)
  peak1.df <- data.frame(peak1_anno) %>%
    dplyr::select(seqnames, start, end, annot.type,annot.symbol,connID)
  
  cicero_peak2 <- str_split(cicero_conns$Peak2, pattern="_", simplify = TRUE) %>%
    as.data.frame()
  colnames(cicero_peak2) <- c("chr","start","end")
  cicero_peak2.gr <- makeGRangesFromDataFrame(cicero_peak2)
  cicero_peak2.gr$connID <- cicero_conns$connID
  peak2_anno <- annotate_regions(cicero_peak2.gr, annotations=anno, ignore.strand=TRUE)
  peak2.df <- data.frame(peak2_anno) %>%
    dplyr::select(seqnames, start, end, annot.type, annot.symbol, connID)
  
  # determine which connections contain an enhancer and a promoter
  peak1_promID <- unique(peak1.df[peak1.df$annot.type == "hg38_genes_promoters",]$connID)
  peak1_enhID <- unique(peak1.df[peak1.df$annot.type == "hg38_enhancers_fantom",]$connID)
  
  peak2_promID <- unique(peak2.df[peak2.df$annot.type == "hg38_genes_promoters",]$connID)
  peak2_enhID <- unique(peak2.df[peak2.df$annot.type == "hg38_enhancers_fantom",]$connID)
  
  # create a list of unique connections with an enhancer and a promoter
  prom1_enh2 <- intersect(peak1_promID, peak2_enhID)
  prom2_enh1 <- intersect(peak2_promID, peak1_enhID)
  prom_enh_connID <- unique(c(prom1_enh2, prom2_enh1))
  
  # filter ccan for connections containing a promoter and an enhancer
  cicero_conns_promEnh <- dplyr::filter(cicero_conns, connID %in% prom_enh_connID)
  
  # swap the peaks so promoter peak is always peak1
  cicero_conns_promEnh <- dplyr::mutate(cicero_conns_promEnh, promPeak = ifelse(connID %in% prom2_enh1, Peak2, Peak1)) %>%
                                          dplyr::mutate(enhPeak = ifelse(connID %in% prom2_enh1, Peak1, Peak2))
  
  # determine the max coaccess score for each promoter
  # coaccess_max <- dplyr::group_by(cicero_conns_promEnh, Peak1) %>% top_n(1, coaccess) %>%
  #   as.data.frame()
  
  coaccess_max <- cicero_conns_promEnh
  
  # annotate the promoter peaks with gene names
  cicero_prom <- str_split(coaccess_max$promPeak, pattern="_", simplify = TRUE) %>%
    as.data.frame()
  colnames(cicero_prom) <- c("chr","start","end")
  cicero_prom.gr <- makeGRangesFromDataFrame(cicero_prom)
  cicero_prom.gr$connID <- coaccess_max$connID
  cicero_prom.gr$coaccess <- coaccess_max$coaccess
  cicero_prom.gr$Peak1 <- coaccess_max$promPeak
  cicero_prom.gr$Peak2 <- coaccess_max$enhPeak

  
  # load the ensembl annotations
  gene.coords <- genes(EnsDb.Hsapiens.v86, filter = ~ gene_biotype == "protein_coding", columns = "gene_name")
  seqlevelsStyle(gene.coords) <- 'UCSC'
  gene.coords <- keepStandardChromosomes(gene.coords, pruning.mode = 'coarse')
  
  # determine which promoter peaks overlap with protein coding genes
  anno.gr <- join_overlap_inner(cicero_prom.gr, gene.coords)
  return(anno.gr)
}

# load atac aggregated object
atacAggr <- readRDS("cellranger_atac_prep/atacAggr_sub97_control.rds")

# # load in annotations available with builtin_annotations()
# # identify promoter peaks
db <- c("hg38_genes_promoters","hg38_enhancers_fantom")
anno <- build_annotations(genome = 'hg38', annotations=db)

anno.ls <- lapply(levels(atacAggr), function(x) {Prom_Enh_Anno(x, anno.db=anno)})
names(anno.ls) <- levels(atacAggr)

# retrieve cell-specific gene expression
celltype <- getSheetNames("analysis_control/deg.celltype.control.xlsx")
cell_deg.ls <- lapply(celltype, function(ident){
  df <- read.xlsx("analysis_control/deg.celltype.control.xlsx", sheet = ident, rowNames = TRUE)
  df$celltype <- ident
  df <- rownames_to_column(df, var = "gene")
  return(df)
})
names(cell_deg.ls) <- c("PCT","PT_VCAM1","PEC","TAL","DCT","DCT2","CNT","PC","ICA","ICB","PODO","ENDO","MES_FIB","FIB","LEUK")

# correlate relative cell-specific gene expression with mean or max prom-enh coaccessibility
Gene_Enh_Coaccess_Exp_Mat <- function(ident) {
  print(paste0("Creating correlation matrix for: ", ident))
  deg.df <- cell_deg.ls[[ident]]

  geneH <- anno.ls[[ident]]

  degs <- unique(deg.df[deg.df$p_val_adj < 0.05, ]$gene)
  print(length(degs))
  
  # iterate over all cell-specific genes and match with mean and max coaccess score
  corr.df <- lapply(degs, function(gene){
    lfc <- deg.df[deg.df$gene == gene,]$avg_logFC
    my.mean <- function(x) ifelse( !all(is.na(x)), mean(x, na.rm=T), NA)
    mean_coaccess <- my.mean(geneH[geneH$gene_name == gene, ]$coaccess)
    my.max <- function(x) ifelse( !all(is.na(x)), max(x, na.rm=T), NA)
    max_coaccess <- my.max(geneH[geneH$gene_name == gene, ]$coaccess)
    Peak1_max <- tryCatch(paste(unique(geneH[geneH$coaccess == max_coaccess, ]$Peak1), 
                          collapse=","), error=function(e) NA)
    Peak2_max <- tryCatch(paste(unique(geneH[geneH$coaccess == max_coaccess, ]$Peak2), 
                                collapse=","), error=function(e) NA)
    df <- data.frame(gene=gene, avg_logFC=lfc, mean_coaccess=mean_coaccess,
                     max_coaccess=max_coaccess, celltype=ident, Peak1_prom=Peak1_max, Peak2_maxEnh=Peak2_max)
    return(df)
  }) %>% bind_rows()
  return(corr.df)
}

corr.df <- lapply(levels(atacAggr), function(ident) {Gene_Enh_Coaccess_Exp_Mat(ident)}) %>%
  bind_rows()
# corr.df <- dplyr::filter(corr.df, avg_logFC > 0)
corr.df <- dplyr::mutate(corr.df, combo = paste0(gene,"_",celltype))

corr.df$celltype <- factor(corr.df$celltype, levels=levels(atacAggr))
ggplot(corr.df, aes(y=avg_logFC, x=mean_coaccess)) + geom_point()
pearson <- cor.test(corr.df$mean_coaccess, corr.df$avg_logFC, method="pearson", conf.level=0.95)
max_cicero <- max(abs(corr.df$mean_coaccess)) * 1.1
max_exp <- max(abs(corr.df$avg_logFC)) * 1.1
p1 <- ggplot(corr.df, aes(x=mean_coaccess, y=avg_logFC)) +
  # geom_smooth(method="lm", color = "darkgray") +
  geom_point(aes(color=celltype)) +
  xlab("Enhancer-Promoter Cicero activity") +
  ylab("Gene expression (avg_logFC)") +
  ggtitle("Cicero Enh-Prom Activity vs. Gene Expression for all celltypes",
          subtitle=paste0("Pearson r^2=",
                          signif(pearson$estimate,2),
                          " pval=",signif(pearson$p.val,2))) +
  xlim(c(0,max_cicero)) +
  ylim(c(-max_exp,max_exp)) +
  theme_minimal() +
  geom_hline(yintercept=0) +
  geom_vline(xintercept=0) 
p1

# calculate a pearson correlation between cicero activity and gene expression for every detected gene
library(ggpmisc)
df.pearson <- lapply(unique(corr.df$gene), function(gene_sel) {
  df <- dplyr::filter(corr.df, gene == gene_sel)
  pearson <- tryCatch(cor.test(df$mean_coaccess, df$avg_logFC,
                               method="pearson", conf.level=0.95),
                      error=function(e) NULL)
  df$corr <- tryCatch(signif(pearson$estimate,2), error=function(e) NULL)
  df$pval <- tryCatch(signif(pearson$p.value,2), error=function(e) NULL)
  df$num_celltypes <- length(unique(df$celltype))
  return(df)
})
df.final <- bind_rows(df.pearson)
df.final <- dplyr::distinct(df.final, combo, .keep_all=TRUE)
write.csv(df.final, file = "analysis_control/ccans/corr_ccan_cicero_mean_coaccess.csv")

# filter for significant correlations
df.sig <- dplyr::filter(df.final, pval < 0.05) %>%
  arrange(-num_celltypes, corr)
library(ggrepel)
gene_sel = "COL4A4"
gene.df <- dplyr::filter(df.sig, gene==gene_sel)
p1 <- ggplot(gene.df, aes(x=mean_coaccess, y=avg_logFC)) +
  geom_smooth(method="lm", color = "darkgray") +
  geom_point(aes(color=celltype)) +
  geom_text_repel(aes(label=celltype)) +
  xlab("Mean Gene-Enhancer Coaccessibility") +
  ylab("Gene expression (avg_logFC)") +
  ggtitle(gene_sel,
          subtitle=paste0("Pearson r^2=",
                          unique(df.sig$corr),
                          " pval=",unique(df.sig$pval))) +
  theme_minimal() +
  geom_hline(yintercept=0) +
  geom_vline(xintercept=0) 
p1

########### do the same for max coaccess
library(ggpmisc)
df.pearson <- lapply(unique(corr.df$gene), function(gene_sel) {
  df <- dplyr::filter(corr.df, gene == gene_sel)
  pearson <- tryCatch(cor.test(df$max_coaccess, df$avg_logFC,
                               method="pearson", conf.level=0.95),
                      error=function(e) NULL)
  df$corr <- tryCatch(signif(pearson$estimate,2), error=function(e) NULL)
  df$pval <- tryCatch(signif(pearson$p.value,2), error=function(e) NULL)
  df$num_celltypes <- length(unique(df$celltype))
  return(df)
})
df.final <- bind_rows(df.pearson)
df.final <- dplyr::distinct(df.final, combo, .keep_all=TRUE)
write.csv(df.final, file = "analysis_control/ccans/corr_ccan_cicero_max_coaccess.csv")


# filter for significant correlations
df.sig <- dplyr::filter(df.final, pval < 0.05) %>%
  arrange(-num_celltypes, corr)
df.sig <- na.omit(df.sig)
library(ggrepel)
gene_sel = "NEDD9"
gene.df <- dplyr::filter(df.sig, gene==gene_sel)
p1 <- ggplot(gene.df, aes(x=max_coaccess, y=avg_logFC)) +
  geom_smooth(method="lm", color = "darkgray") +
  geom_point(aes(color=celltype)) +
  geom_text_repel(aes(label=celltype)) +
  xlab("Max Gene-Enhancer Coaccessibility") +
  ylab("Gene expression (avg_logFC)") +
  ggtitle(gene_sel,
          subtitle=paste0("Pearson r^2=",
                          unique(df.sig$corr),
                          " pval=",unique(df.sig$pval))) +
  theme_minimal() +
  geom_hline(yintercept=0) +
  geom_vline(xintercept=0) 
p1



                  